File: backend/app/schemas/plan.py
"""Request and response schemas for plan endpoints."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


Industry = Literal[
    "Healthcare",
    "Finance",
    "Education",
    "Hotel",
    "E-commerce",
    "AI",
    "SaaS",
    "Other",
]
Budget = Literal["Low", "Medium", "High"]
ProjectType = Literal["MVP", "Startup", "Enterprise", "Student Project"]
ExperienceLevel = Literal["Beginner", "Intermediate", "Advanced"]


class GeneratePlanRequest(BaseModel):
    """Input payload for plan generation."""

    project_idea: str = Field(..., min_length=10, max_length=5000)
    industry: Industry = "Other"
    budget: Budget = "Medium"
    project_type: ProjectType = "MVP"
    experience_level: ExperienceLevel = "Intermediate"

    @field_validator("project_idea")
    @classmethod
    def strip_idea(cls, value: str) -> str:
        cleaned = value.strip()
        if len(cleaned) < 10:
            raise ValueError("Project idea must be at least 10 characters.")
        return cleaned


class GeneratePlanResponse(BaseModel):
    """Generated plan markdown."""

    plan: str


class SavePlanRequest(BaseModel):
    """Payload to persist a generated plan."""

    project_idea: str = Field(..., min_length=10, max_length=5000)
    industry: Industry = "Other"
    budget: Budget = "Medium"
    project_type: ProjectType = "MVP"
    experience_level: ExperienceLevel = "Intermediate"
    generated_plan: str = Field(..., min_length=50)


class SavePlanResponse(BaseModel):
    """Response after saving a plan."""

    id: int
    message: str = "Plan saved successfully."


class PlanHistoryItem(BaseModel):
    """Single plan record from history."""

    id: int
    project_idea: str
    industry: str
    budget: str
    project_type: str
    experience_level: str
    generated_plan: str
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/plan.ts
export type Industry =
  | "Healthcare"
  | "Finance"
  | "Education"
  | "Hotel"
  | "E-commerce"
  | "AI"
  | "SaaS"
  | "Other";

export type Budget = "Low" | "Medium" | "High";
export type ProjectType = "MVP" | "Startup" | "Enterprise" | "Student Project";
export type ExperienceLevel = "Beginner" | "Intermediate" | "Advanced";

export interface GeneratePlanRequest {
  project_idea: string;
  industry: Industry;
  budget: Budget;
  project_type: ProjectType;
  experience_level: ExperienceLevel;
}

export interface GeneratePlanResponse {
  plan: string;
}

export interface PlanHistoryItem {
  id: number;
  project_idea: string;
  industry: string;
  budget: string;
  project_type: string;
  experience_level: string;
  generated_plan: string;
  created_at: string;
}

export interface SavePlanRequest extends GeneratePlanRequest {
  generated_plan: string;
}

export interface SavePlanResponse {
  id: number;
  message: string;
}

export const INDUSTRIES: Industry[] = [
  "Healthcare",
  "Finance",
  "Education",
  "Hotel",
  "E-commerce",
  "AI",
  "SaaS",
  "Other",
];

export const BUDGETS: Budget[] = ["Low", "Medium", "High"];
export const PROJECT_TYPES: ProjectType[] = [
  "MVP",
  "Startup",
  "Enterprise",
  "Student Project",
];
export const EXPERIENCE_LEVELS: ExperienceLevel[] = [
  "Beginner",
  "Intermediate",
  "Advanced",
];
